sdk/core/azure-core/src/main/java/com/azure/core/implementation/util/LruCache.java

```diff
@@ -0,0 +1,141 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+package com.azure.core.implementation.util;
+
+import com.azure.core.util.logging.ClientLogger;
+
+import java.util.HashMap;
+import java.util.LinkedList;
+import java.util.Map;
+import java.util.Queue;
+import java.util.function.Function;
+
+/**
+ * A least recently used cache that will retain the most recently used key-values until a cache size limit is reached.
+ * <p>
+ * The least recently used cache will eject values that have been added or accessed least recently, offering the ability
+ * to limit the size of the cache while still receiving the benefit of retaining the most recently used values. The
+ * cache does have limitations where it is possible for a most commonly used value to be ejected due to a streak of less
+ * commonly but more recently used values being added, which won't be covered in this implementation.
+ */
+public final class LruCache<Key, Value> {
+    private static final ClientLogger LOGGER = new ClientLogger(LruCache.class);
+
+    private final int cacheSizeLimit;
+
+    // Maintains the key-value pairs in the cache.
+    private final Map<Key, Value> cache;
+
+    // Maintains the least recently used ordering of keys.
+    private final Queue<Key> lruKeys;
+
+    /**
+     * Creates an {@link LruCache} instance that caches the {@code cacheSize} most recently used values.
+     *
+     * @param cacheSizeLimit The number of values that can be cached at any time.
+     * @throws IllegalArgumentException If {@code cacheSize} is less than or equal to 0.
+     */
+    public LruCache(int cacheSizeLimit) {
+        if (cacheSizeLimit <= 0) {
+            throw LOGGER.logExceptionAsError(
+                new IllegalArgumentException("'cacheSizeLimit' cannot be less than or equal to 0."));
+        }
+
+        this.cacheSizeLimit = cacheSizeLimit;
+
+        // Given that all operations on the LRU cache are synchronized the implementation types can be non-thread safe.
+        this.cache = new HashMap<>();
+        this.lruKeys = new LinkedList<>();
```

Comment:
We should use [ConcurrentHashMap](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ConcurrentHashMap.html) and [ConcurrentLinkedQueue](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ConcurrentLinkedQueue.html) here.